File: version_js/test/test_endTurn.ts
// @ts-ignore
import {CaseType, Map, PRODUCTION_MODE} from "../src/model/map"
import {Box, Cardinal, Machine, MachineStuff} from "../src/model/machine";
import {describe, it} from "mocha";
import {Difficulty, Level} from "../src/model/difficulty";
import {Config} from "../src/utils/config";
import {Vector2D} from "../src/utils/utilities";

let assert = require('assert');

describe('End Turn', function() {
    let test_map: Map;
    let test_gate : Vector2D;
    let test_sources : Array<Vector2D>;

    function firstCaseEmpty() {
        for (let i = 0; i < test_map.getWidth; i++) {
            for (let j = 0; j < test_map.getHeight; j++) {
                if (test_map.getCase(i,j)?.isEmpty) {
                    return new Vector2D(i,j);
                }
            }
        }

        return null;
    }

    beforeEach('Prepare', function () {
        test_map = new Map(Difficulty.getDifficultyByID(Level.EASY));

        test_sources = new Array<Vector2D>();
        for (let ligne = 0; ligne < test_map.getWidth; ligne++) {
            for (let colonne = 0; colonne < test_map.getHeight; colonne++) {
                let currentCase = test_map.getCase(ligne, colonne);
                if (currentCase != null) {
                    if (currentCase.caseType == CaseType.CASE_SOURCE) {
                        test_sources.push(new Vector2D(ligne, colonne));
                    } else if (currentCase.caseType == CaseType.CASE_GATE) {
                        test_gate = new Vector2D(ligne, colonne);
                    }
                }
            }
        }
    });

    it('Production Fise', function () {
        // Assumption that the creation of a map has been tested beforehand
        // Create map
        let map = new Map(Difficulty.getDifficultyByID(Level.EASY));

        let numberE = map.getNumberE;
        let numberDD = map.getNumberDD;
        let numberFise = map.getNumberFISE;

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        map.productionFise();

        // Checking values
        assert.equal(map.getNumberE, numberE + (Config.constants.PRODUCTION_FISE_E * numberFise));
        assert.equal(map.getNumberDD, numberDD + (Config.constants.PRODUCTION_FISE_DD * numberFise));
    });

    it('Production Fisa', function () {
        // Assumption that the creation of a map has been tested beforehand
        // Create map
        let map = new Map(Difficulty.getDifficultyByID(Level.EASY));

        let numberE = map.getNumberE;
        let numberDD = map.getNumberDD;
        let numberFisa = map.getNumberFISA;

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        map.productionFisa();

        // Checking values
        assert.equal(map.getNumberE, numberE);
        assert.equal(map.getNumberDD, numberDD);

        map.addTurn(Config.constants.NB_TURN_FISA - 1);
        assert.equal(map.getNumberTurn, Config.constants.NB_TURN_FISA);

        // Calls the function to test
        map.productionFisa();

        // Checking values
        assert.equal(map.getNumberE, numberE + (Config.constants.PRODUCTION_FISA_E * numberFisa));
        assert.equal(map.getNumberDD, numberDD);

        // Recovery of value before purchase
        numberE = map.getNumberE;
        numberDD = map.getNumberDD;

        map.addTurn(Config.constants.NB_TURN_FISA);
        map.changeProductionFISA();

        // Calls the function to test
        map.productionFisa();

        // Checking values
        assert.equal(map.getNumberE, numberE);
        assert.equal(map.getNumberDD, numberDD + (Config.constants.PRODUCTION_FISA_DD * numberFisa));
    });

    it('Move box', function () {
        // Assumption that the operation of a machine, case and box has been tested beforehand
        // Machine creation
        let numberR = 10;

        let beltCase1 = firstCaseEmpty();
        let outCase1 : Vector2D;

        let machine1 = new Machine(MachineStuff.MS_CONVEYOR_BELT);
        let machine2 = new Machine(MachineStuff.MS_CONVEYOR_BELT);

        let cardinalIn : Cardinal;
        let cardinalOut : Cardinal;
        // @ts-ignore
        if (test_map.getCase(beltCase1.x - 1, beltCase1.y)?.isEmpty) {
            machine1.rotate(1);
            machine2.rotate(1);
            cardinalOut = Cardinal.WEST;
            cardinalIn = Cardinal.EAST;
            // @ts-ignore
            outCase1 = new Vector2D(beltCase1.x - 1, beltCase1.y);
        } else { // @ts-ignore
            if (test_map.getCase(beltCase1.x, beltCase1.y - 1)?.isEmpty) {
                machine1.rotate(2);
                machine2.rotate(2);
                cardinalOut = Cardinal.NORTH;
                cardinalIn = Cardinal.SOUTH;
                // @ts-ignore
                outCase1 = new Vector2D(beltCase1.x, beltCase1.y - 1);
            } else { // @ts-ignore
                if (test_map.getCase(beltCase1.x + 1, beltCase1.y)?.isEmpty) {
                    machine1.rotate(3);
                    machine2.rotate(3);
                    cardinalOut = Cardinal.EAST;
                    cardinalIn = Cardinal.WEST;
                    // @ts-ignore
                    outCase1 = new Vector2D(beltCase1.x + 1, beltCase1.y);
                } else {
                    // @ts-ignore
                    outCase1 = new Vector2D(beltCase1.x, beltCase1.y + 1);
                    cardinalOut = Cardinal.SOUTH;
                    cardinalIn = Cardinal.NORTH;
                }
            }
        }

        let box1 = new Box(numberR,0);
        machine1.setBox(cardinalOut, box1);

        // @ts-ignore
        let ca1 = test_map.getCase(beltCase1.x, beltCase1.y);
        ca1?.setMachine(machine1);

        let ca2 = test_map.getCase(outCase1.x, outCase1.y);
        ca2?.setMachine(machine2);

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.moveResources();

        // Checking values
        assert.equal(machine1.getBox(cardinalOut), null);
        let box1R = machine2.getBox(cardinalIn);
        assert.notEqual(box1R, null);
        assert.equal(box1R?.numberResources, numberR);
        assert.equal(box1R?.numberGarbage, 0);
    });

    it('Generate Ressource', function () {
        let source1 = test_map.getCase(test_sources[0].x, test_sources[0].y);
        assert.equal(source1?.caseType, CaseType.CASE_SOURCE);
        let source2 = test_map.getCase(test_sources[1].x, test_sources[1].y);
        assert.equal(source2?.caseType, CaseType.CASE_SOURCE);

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.generateResources();

        let boxSource1 = source1?.getBox();
        assert.equal(boxSource1, null);


        let boxSource2 = source2?.getBox();
        assert.equal(boxSource2, null);

        test_map.addTurn(Config.constants.NB_TURN_PRODUCTION_SOURCE-1);

        // Calls the function to test
        test_map.generateResources();

        // Checking values
        boxSource1 = source1?.getBox();
        assert.notEqual(boxSource1, null);
        assert.equal(boxSource1?.numberGarbage, 0);
        assert.equal(boxSource1?.numberResources, Config.constants.NB_RESOURCE_PRODUCT_BY_SOURCE);

        boxSource2 = source2?.getBox();
        assert.notEqual(boxSource2, null);
        assert.equal(boxSource2?.numberGarbage, 0);
        assert.equal(boxSource2?.numberResources, Config.constants.NB_RESOURCE_PRODUCT_BY_SOURCE);
    });

    it('Generate Garbage', function () {
        let numberR = 10;
        // Assumption that the operation of a machine, case and box has been tested beforehand
        // Machine creation
        let gate = test_map.getCase(test_gate.x, test_gate.y);
        let box = new Box(numberR, 0);
        gate?.setBox(box);

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.generateGarbage();

        // Checking values
        let boxGateR = gate?.getBox();
        assert.notEqual(boxGateR, null);
        assert.equal(boxGateR?.numberGarbage, numberR);
        assert.equal(boxGateR?.numberResources, 0);
    });

    it('Recycling Center', function () {
        // Assumption that the operation of a machine, case and box has been tested beforehand
        // Machine creation
        let cEmpty = firstCaseEmpty();
        // @ts-ignore
        let ca = test_map.getCase(cEmpty?.x, cEmpty?.y);
        let machine = new Machine(MachineStuff.MS_RECYCLING_CENTER);
        let box = new Box(0, Config.constants.NUMBER_WASTE_TO_PRODUCT_RESOURCE);
        machine.setBox(Cardinal.SOUTH, box);
        ca?.setMachine(machine);

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.activateRecyclingCenters();

        // Checking values
        let boxOut = machine.getBox(Cardinal.SOUTH);
        // @ts-ignore
        assert.equal(boxOut.numberGarbage, 0);
        // @ts-ignore
        assert.equal(boxOut.numberResources, 0);
        // @ts-ignore
        let out = test_map.getCase(cEmpty?.x, cEmpty?.y + 1);
        let boxR = out?.getBox();
        assert.notEqual(boxR, undefined);
        // @ts-ignore
        assert.equal(boxR.numberGarbage, 0);
        // @ts-ignore
        assert.equal(boxR.numberResources, 1);
    });

    it('Activate Collector', function () {
        let numberR = 3;
        // Assumption that the operation of a machine, case and box has been tested beforehand
        // Machine creation
        let ca1 = test_map.getCase(test_sources[0].x, test_sources[0].y);
        let box = new Box(numberR,0);
        ca1?.setBox(box);

        let collectorCase : Vector2D;
        if (test_map.getCase(test_sources[0].x, test_sources[0].y + 1)?.isEmpty) {
            collectorCase = new Vector2D(test_sources[0].x, test_sources[0].y + 1);
        } else if (test_map.getCase(test_sources[0].x, test_sources[0].y - 1)?.isEmpty) {
            collectorCase = new Vector2D(test_sources[0].x, test_sources[0].y - 1);
        } else if (test_map.getCase(test_sources[0].x + 1, test_sources[0].y)?.isEmpty) {
            collectorCase = new Vector2D(test_sources[0].x + 1, test_sources[0].y);
        } else {
            collectorCase = new Vector2D(test_sources[0].x - 1, test_sources[0].y);
        }

        let ca2 = test_map.getCase(collectorCase.x, collectorCase.y);
        let machine = new Machine(MachineStuff.MS_COLLECTOR);
        ca2?.setMachine(machine);

        let outCase : Vector2D;
        if (test_map.getCase(collectorCase.x - 1, collectorCase.y)?.isEmpty) {
            machine.rotate(1);
            outCase = new Vector2D(collectorCase.x - 1, collectorCase.y);
        } else if (test_map.getCase(collectorCase.x, collectorCase.y - 1)?.isEmpty) {
            machine.rotate(2);
            outCase = new Vector2D(collectorCase.x, collectorCase.y - 1);
        } else if (test_map.getCase(collectorCase.x + 1, collectorCase.y)?.isEmpty) {
            machine.rotate(3);
            outCase = new Vector2D(collectorCase.x + 1, collectorCase.y);
        } else {
            outCase = new Vector2D(collectorCase.x, collectorCase.y + 1);
        }

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.activateCollectors();

        // Checking values
        assert.equal(box.numberGarbage, 0);
        assert.equal(box.numberResources, numberR-1);
        let boxR = test_map.getCase(outCase.x, outCase.y)?.getBox();
        assert.notEqual(boxR, null);
        assert.equal(boxR?.numberGarbage, 0);
        assert.equal(boxR?.numberResources, 1);
    });

    it('Reset Ressource Garbage', function () {
        let type = MachineStuff.MS_CONVEYOR_BELT;

        let numberR = 10;
        let numberG = 10;

        // Assumption that the operation of a machine, case and box has been tested beforehand
        // Machine creation
        let emptyCase1 = firstCaseEmpty();
        // @ts-ignore
        let ca1 = test_map.getCase(emptyCase1.x, emptyCase1.y);
        let machine: Machine | undefined = new Machine(type);
        ca1?.setMachine(machine);
        let boxM1 = new Box(numberR,numberG);
        machine.setBox(Cardinal.NORTH, boxM1);
        let boxM2 = new Box(numberR,numberG);
        machine.setBox(Cardinal.SOUTH, boxM2);
        let boxM3 = new Box(numberR,numberG);
        machine.setBox(Cardinal.EAST, boxM3);
        let boxM4 = new Box(numberR,numberG);
        machine.setBox(Cardinal.WEST, boxM4);

        let emptyCase2 = firstCaseEmpty();
        // @ts-ignore
        let ca2 = test_map.getCase(emptyCase2.x, emptyCase2.y);
        let boxS = new Box(numberR,numberG);
        ca2?.setBox(boxS);

        // This function uses functions of another implemented type.
        // Each of the functions has been tested in the previous tests.
        // Calls the function to test
        test_map.resetResourcesGarbage();

        // Checking values
        machine = ca1?.getMachine();
        assert.equal(machine?.getBox(Cardinal.NORTH), boxM1);
        assert.equal(machine?.getBox(Cardinal.SOUTH), boxM2);
        assert.equal(machine?.getBox(Cardinal.EAST), boxM3);
        assert.equal(machine?.getBox(Cardinal.WEST), boxM4);
        assert.equal(ca2?.getBox(), undefined);

        let gate = test_map.getCase(test_gate.x, test_gate.y);
        assert.notEqual(gate, null);
        let boxGate = gate?.getBox();
        assert.notEqual(boxGate, undefined);
        assert.equal(boxGate?.numberGarbage, numberG);
        assert.equal(boxGate?.numberResources, 0);
    });
});

File: version_c/tests/test_endTurn.c
/*!
 * \author Valentin DREANO
 * \version 0.1
 * \date 17/04/2021
 *
 * Contains all the suite for endTurn function tests
 * The assumption is made that the setters and getter are functional
 *
 */

#include "test_endTurn.h"

Vector2D testEndTurn_gate;
Vector2D testEndTurn_source1;
Vector2D testEndTurn_source2;

test_List testEndTurn_listTests[TESTENDTURN_NUMBERTESTS] = {
        {"EndTurn_Production_Fise\0", testEndTurn_productionFise},
        {"EndTurn_Production_Fisa\0", testEndTurn_productionFisa},
        {"EndTurn_Move_Resource\0", testEndTurn_moveResource},
        {"EndTurn_Generate_Resource\0", testEndTurn_generateResource},
        {"EndTurn_Generate_Garbage\0", testEndTurn_generateGarbage},
        {"EndTurn_RecyclingCenters\0", testEndTurn_RecyclingCenters},
        {"EndTurn_Collector\0", testEndTurn_Collector},
        {"EndTurn_Production_resetResourceGarbage\0", testEndTurn_resetResourceGarbage}
};

int testEndTurn_createSuite(CU_pSuite pSuite) {

    /* add a suite to the registry */
    pSuite = CU_add_suite("Test_Function_EndTurn", testEndTurn_initSuite, testEndTurn_cleanSuite);
    if (NULL == pSuite) {
        return EXIT_FAILURE;
    }

    /* add the tests to the suite */
    for (int i = 0; i < TESTENDTURN_NUMBERTESTS; i++) {
        if (NULL == CU_add_test(pSuite, testEndTurn_listTests[i].description, testEndTurn_listTests[i].function)) {
            return EXIT_FAILURE;
        }
    }

    return EXIT_SUCCESS;
}

int testEndTurn_initSuite() {
    testEndTurn_gate.x = 3;
    testEndTurn_gate.y = 6;
    testEndTurn_source1.x = 3;
    testEndTurn_source1.y = 5;
    testEndTurn_source2.x = 7;
    testEndTurn_source2.y = 5;
    return 0;
}

int testEndTurn_cleanSuite() {
    return 0;
}

void testEndTurn_productionFise() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    Map *map = map_create(DIFFICULTY_EASY);

    int numberE = map_getNumberE(map);
    int numberDD = map_getNumberDD(map);
    int numberFise = map_getNumberFISE(map);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_productionFise(map);

    // Checking values
    CU_ASSERT_EQUAL(map_getNumberE(map), numberE + (PRODUCTION_FISE_E * numberFise));
    CU_ASSERT_EQUAL(map_getNumberDD(map), numberDD + (PRODUCTION_FISE_DD * numberFise));

    // Destroy map
    map_destroy(map);
}

void testEndTurn_productionFisa() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    Map *map = map_create(DIFFICULTY_EASY);

    // Recovery of value before purchase
    int numberE = map_getNumberE(map);
    int numberDD = map_getNumberDD(map);
    int numberFisa = map_getNumberFISA(map);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_productionFisa(map);

    // Checking values
    CU_ASSERT_EQUAL(map_getNumberE(map), numberE);
    CU_ASSERT_EQUAL(map_getNumberDD(map), numberDD);

    map_setNumberTurn(map, NB_TURN_FISA - 1);

    // Calls the function to test
    map_utils_productionFisa(map);

    // Checking values
    CU_ASSERT_EQUAL(map_getNumberE(map), numberE + (PRODUCTION_FISA_E * numberFisa));
    CU_ASSERT_EQUAL(map_getNumberDD(map), numberDD);


    // Recovery of value before purchase
    numberE = map_getNumberE(map);
    numberDD = map_getNumberDD(map);

    map_setNumberTurn(map, NB_TURN_FISA);
    map_changeProductionFISA(map);

    // Calls the function to test
    map_utils_productionFisa(map);

    // Checking values
    CU_ASSERT_EQUAL(map_getNumberE(map), numberE);
    CU_ASSERT_EQUAL(map_getNumberDD(map), numberDD + (PRODUCTION_FISA_DD * numberFisa));

    // Destroy map
    map_destroy(map);
}

void testEndTurn_moveResource() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    Map *map = map_create(DIFFICULTY_EASY);

    // Assumption that the operation of a machine, case and box has been tested beforehand
    // Machine creation
    int numberR = 10;
    int numberG = 10;
    Case *ca1 = map_getCase(0, 0, map);
    Machine *machine1 = machine_create(MS_CONVEYOR_BELT);
    Box *box1 = box_create(numberR,0);
    machine_addBox(machine1, SOUTH, box1);
    case_addMachine(ca1, machine1);

    Case *ca2 = map_getCase(0, 1, map);
    Machine *machine2 = machine_create(MS_CONVEYOR_BELT);
    case_addMachine(ca2, machine2);

    Case *ca3 = map_getCase(1, 0, map);
    Machine *machine3 = machine_create(MS_CONVEYOR_BELT);
    Box *box2 = box_create(0,numberG);
    machine_addBox(machine3, SOUTH, box2);
    case_addMachine(ca3, machine3);

    Case *ca4 = map_getCase(1, 1, map);
    Machine *machine4 = machine_create(MS_JUNKYARD);
    case_addMachine(ca4, machine4);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_moveResources(map);
    // Checking values
    CU_ASSERT_EQUAL(machine_getBox(machine1,SOUTH), NULL);
    CU_ASSERT_EQUAL(machine_getBox(machine3,SOUTH), NULL);
    Box* box1R = machine_getBox(machine2,NORTH);
    CU_ASSERT_EQUAL(box_getNumberResource(box1R), numberR);
    CU_ASSERT_EQUAL(box_getNumberGarbage(box1R), 0);;
    Box* box2R = machine_getBox(machine4,NORTH);
    CU_ASSERT_EQUAL(box_getNumberResource(box2R), 0);
    CU_ASSERT_EQUAL(box_getNumberGarbage(box2R), numberG);

    // Destroy map
    map_destroy(map);
}

void testEndTurn_generateResource() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    srand(1); // NOLINT(cert-msc51-cpp)
    Map *map = map_create(DIFFICULTY_EASY);

    Case *source1 = map_getCase(testEndTurn_source1.x, testEndTurn_source1.y, map);
    CU_ASSERT_EQUAL_FATAL(case_getType(source1), CASE_SOURCE);
    Case *source2 = map_getCase(testEndTurn_source2.x, testEndTurn_source2.y, map);
    CU_ASSERT_EQUAL_FATAL(case_getType(source2), CASE_SOURCE);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_generateResources(map);

    Box *boxSource1 = case_getBox(source1);
    CU_ASSERT_EQUAL_FATAL(boxSource1, NULL);

    Box *boxSource2 = case_getBox(source2);
    CU_ASSERT_EQUAL_FATAL(boxSource2, NULL);

    map_setNumberTurn(map, 9);

    // Calls the function to test
    map_utils_generateResources(map);

    // Checking values
    boxSource1 = case_getBox(source1);
    CU_ASSERT_NOT_EQUAL_FATAL(boxSource1, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxSource1), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(boxSource1), NB_RESOURCE_PRODUCT_BY_SOURCE);

    boxSource2 = case_getBox(source2);
    CU_ASSERT_NOT_EQUAL_FATAL(boxSource2, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxSource2), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(boxSource2), NB_RESOURCE_PRODUCT_BY_SOURCE);

    // Destroy map
    map_destroy(map);
}

void testEndTurn_generateGarbage() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    srand(1); // NOLINT(cert-msc51-cpp)
    Map *map = map_create(DIFFICULTY_EASY);

    int numberR = 10;
    // Assumption that the operation of a machine, case and box has been tested beforehand
    // Machine creation
    Case *gate = map_getCase(testEndTurn_gate.x, testEndTurn_gate.y, map);
    Box *box = box_create(numberR, 0);
    case_addBox(gate, box);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_generateGarbage(map);

    // Checking values
    Box *boxGateR = case_getBox(gate);
    CU_ASSERT_NOT_EQUAL_FATAL(boxGateR, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxGateR), numberR);
    CU_ASSERT_EQUAL(box_getNumberResource(boxGateR), 0);

    // Destroy map
    map_destroy(map);
}

void testEndTurn_RecyclingCenters() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    Map *map = map_create(DIFFICULTY_EASY);

    // Assumption that the operation of a machine, case and box has been tested beforehand
    // Machine creation
    Case *ca = map_getCase(0, 0, map);
    Machine *machine = machine_create(MS_RECYCLING_CENTER);
    Box *box = box_create(0, NUMBER_WASTE_TO_PRODUCT_RESOURCE);
    machine_addBox(machine, SOUTH, box);
    case_addMachine(ca, machine);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_activateRecyclingCenters(map);

    // Checking values
    CU_ASSERT_EQUAL(box_getNumberGarbage(box), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(box), 0);
    Box *boxR = case_getBox(map_getCase(0,1,map));
    CU_ASSERT_NOT_EQUAL_FATAL(boxR, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxR), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(boxR), 1);

    // Destroy map
    map_destroy(map);
}

void testEndTurn_Collector() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    srand(1); // NOLINT(cert-msc51-cpp)
    Map *map = map_create(DIFFICULTY_EASY);

    int numberR = 3;
    // Assumption that the operation of a machine, case and box has been tested beforehand
    // Machine creation
    Case *ca1 = map_getCase(3, 5, map);
    Box*box = box_create(numberR,0);
    case_addBox(ca1, box);

    Case *ca2 = map_getCase(4, 5, map);
    Machine *machine = machine_create(MS_COLLECTOR);
    case_addMachine(ca2, machine);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_activateCollectors(map);

    // Checking values
    CU_ASSERT_EQUAL(box_getNumberGarbage(box), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(box), numberR-1);
    Box *boxR = case_getBox(map_getCase(4,6,map));
    CU_ASSERT_NOT_EQUAL_FATAL(boxR, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxR), 0);
    CU_ASSERT_EQUAL(box_getNumberResource(boxR), 1);

    // Destroy map
    map_destroy(map);
}

void testEndTurn_resetResourceGarbage() {
    // Assumption that the creation of a map has been tested beforehand
    // Create map
    srand(1); // NOLINT(cert-msc51-cpp)
    Map *map = map_create(DIFFICULTY_EASY);

    MachineStuff type = MS_CONVEYOR_BELT;

    int numberR = 10;
    int numberG = 10;

    // Assumption that the operation of a machine, case and box has been tested beforehand
    // Machine creation
    Case *ca1 = map_getCase(0, 0, map);
    Machine *machine = machine_create(type);
    Box *boxM1 = box_create(numberR,numberG);
    machine_addBox(machine, NORTH, boxM1);
    Box *boxM2 = box_create(numberR,numberG);
    machine_addBox(machine, SOUTH, boxM2);
    Box *boxM3 = box_create(numberR,numberG);
    machine_addBox(machine, EAST, boxM3);
    Box *boxM4 = box_create(numberR,numberG);
    machine_addBox(machine, WEST, boxM4);
    case_addMachine(ca1, machine);

    Case *ca2 = map_getCase(0, 1, map);
    Box *boxS = box_create(numberR,numberG);
    case_addBox(ca2, boxS);

    // This function uses functions of another implemented type.
    // Each of the functions has been tested in the previous tests.
    // Calls the function to test
    map_utils_resetResourcesGarbage(map);

    // Checking values
    CU_ASSERT_EQUAL(machine_getBox(machine, NORTH), boxM1);
    CU_ASSERT_EQUAL(machine_getBox(machine, SOUTH), boxM2);
    CU_ASSERT_EQUAL(machine_getBox(machine, EAST), boxM3);
    CU_ASSERT_EQUAL(machine_getBox(machine, WEST), boxM4);
    CU_ASSERT_EQUAL(case_getBox(ca2), NULL);

    Case *gate = map_getCase(testEndTurn_gate.x, testEndTurn_gate.y, map);
    CU_ASSERT_NOT_EQUAL_FATAL(gate, NULL);
    Box *boxGate = case_getBox(gate);
    CU_ASSERT_NOT_EQUAL_FATAL(boxGate, NULL);
    CU_ASSERT_EQUAL(box_getNumberGarbage(boxGate), numberG);
    CU_ASSERT_EQUAL(box_getNumberResource(boxGate), 0);

    // Destroy map
    map_destroy(map);
}